### ISSUE TITLE
k3s crictl command not working

### ISSUE THREAD

**Issue by stmaute** [k3s-io/k3s#2891]
**Environmental Info:**
K3s Version: 
k3s -v
k3s version v1.20.2+k3s1 (1d4adb03)
go version go1.15.5

Node(s) CPU architecture, OS, and Version: 
name -a
Linux ip-10-1-51-95.eu-central-1.compute.internal 4.14.214-160.339.amzn2.x86_64 #1 SMP Sun Jan 10 05:53:05 UTC 2021 x86_64 x86_64 x86_64 GNU/Linu

Cluster Configuration:
Single node cluster

**Describe the bug:**
The embedded crictl command is not working because of no such file or directory.

k3s crictl ps
FATA[2021-02-03T11:53:48.714116534Z] load config file: stat /var/lib/rancher/k3s/data/8f4b194129852507eab4a55117fc942e0688ec9a70ffdaa5911ccc6652220f76/bin/crictl.yaml: no such file or directory

Generall question:
How should the k3s crictl command be executed?

**Comment by brandond** [comment]
Are you running crictl before starting k3s? It should drop a config file on first run.

**Comment by stmaute** [comment]
@brandond  Thanks for the quick response.
I run the command after installing k3s with the one liner: _curl -sfL https://get.k3s.io | sh -_
Yesterday I did some digging an found the crictl.yaml file in /var/lib/rancher/k3s/agent/etc/crictl.yaml.
But the command uses the following path: /var/lib/rancher/k3s/data/8f4b194129852507eab4a55117fc942e0688ec9a70ffdaa5911ccc6652220f76/bin/crictl.yam

With the previous version "v1.19.7+k3s1" I got it working after changing some directory ownership.

**Comment by brandond** [comment]
It will fall back to that file, and throw an error if it can't find it, as a last resort if it can't read several other files first.  Are you running crictl as root or with sudo? If not, does your account have access to /var/lib/rancher/k3s/agent/etc/crictl.yaml?

**Comment by stmaute** [comment]
I don't run it with sudo. I have change the owner ship to my user.
I have access to  /var/lib/rancher/k3s/agent/etc/crictl.yaml and if I explicitly use the config file with the -c parameter the command succeeds.
k3s crictl -c /var/lib/rancher/k3s/agent/etc/crictl.yaml ps

**Comment by brandond** [comment]
Ah, it appears this was broken by 3c7fd3d37b079b512093d9e810c3e6cc9e002c2a. While this offers the correct behavior for rootless execution, we don't want to skip checking the default paths in other circumstances.

In the mean time, using sudo should offer an easy workaround.

**Comment by brandond** [comment]
It's also broken in other ways, since `${HOME}` shell expansion isn't occurring:
```
[sysadm@centos01 ~]$ strace -qq -f -e trace=open,stat,newfstatat -e signal=TERM k3s crictl ps
[pid 18242] newfstatat(AT_FDCWD, "k3s", {st_mode=S_IFREG|0755, st_size=59613184, ...}, 0) = 0
[pid 18242] newfstatat(AT_FDCWD, "/var/lib/rancher/k3s/data/e5a08dc33f8f00922b1cd23dbe6cd6b8e25205ebb6d3a9898fc829e926ba71d4", {st_mode=S_IFDIR|0755, st_size=28, ...}, 0) = 0
[pid 18242] newfstatat(AT_FDCWD, "/var/lib/rancher/k3s/data/e5a08dc33f8f00922b1cd23dbe6cd6b8e25205ebb6d3a9898fc829e926ba71d4/bin/crictl", {st_mode=S_IFREG|0755, st_size=113305864, ...}, 0) = 0
[pid 18242] newfstatat(AT_FDCWD, "/proc", {st_mode=S_IFDIR|0555, st_size=0, ...}, 0) = 0
[pid 18242] newfstatat(AT_FDCWD, "/proc", {st_mode=S_IFDIR|0555, st_size=0, ...}, 0) = 0
[pid 18242] newfstatat(AT_FDCWD, "/sys/kernel/mm/hugepages/hugepages-2048kB", {st_mode=S_IFDIR|0755, st_size=0, ...}, AT_SYMLINK_NOFOLLOW) = 0
[pid 18242] newfstatat(AT_FDCWD, "/sys/kernel/mm/hugepages/hugepages-1048576kB", {st_mode=S_IFDIR|0755, st_size=0, ...}, AT_SYMLINK_NOFOLLOW) = 0
[pid 18242] newfstatat(AT_FDCWD, "crictl", 0xc00087fa38, 0) = -1 ENOENT (No such file or directory)
[pid 18242] newfstatat(AT_FDCWD, "${HOME}/.rancher/k3s/agent/etc/crictl.yaml", 0xc000d8a5e8, 0) = -1 ENOENT (No such file or directory)
[pid 18242] newfstatat(AT_FDCWD, "/var/lib/rancher/k3s/data/e5a08dc33f8f00922b1cd23dbe6cd6b8e25205ebb6d3a9898fc829e926ba71d4/bin/crictl.yaml", 0xc000d8a6b8, 0) = -1 ENOENT (No such file or directory)
FATA[2021-02-04T11:24:15.372573616-08:00] load config file: stat /var/lib/rancher/k3s/data/e5a08dc33f8f00922b1cd23dbe6cd6b8e25205ebb6d3a9898fc829e926ba71d4/bin/crictl.yaml: no such file or directory
```

**Comment by stmaute** [comment]
Very cool you already start to fix this. Thx!
In the meanwhile I have found a workaround.

**Comment by rancher-max** [comment]
This is working on commit `6e768c30`. There is another file that the user needs permission to or else you will get this error:
`FATA[2021-02-17T17:12:09.835265344Z] connect: connect endpoint 'unix:///run/k3s/containerd/containerd.sock', make sure you are running as root and the endpoint has been started: context deadline exceeded`.

For simplicity's sake, I ran the following after installing k3s:
```
$ sudo chmod 777 /var/lib/rancher/k3s/ -R
$ sudo chmod 777 /run/k3s/containerd/ -R

```

**Comment by xclydes** [comment]
I was experiencing this issue in `k3s version v1.20.2+k3s1 (1d4adb03)`, however I was able to work around it by manually specifying the location of the `crictl.yaml` on the command line.
All the commands below have to be run as `root`.
```
root@k3s-worker-2:~# find / -name "crictl.yaml"
9372-8938c33dd1eb/mount’: Input/output error
/mnt/storage1/k3s/agent/etc/crictl.yaml
```
With the path from above, executing crictl command worked.
```
root@k3s-worker-2:~# crictl --config /mnt/storage1/k3s/agent/etc/crictl.yaml  ps
CONTAINER           IMAGE               CREATED             STATE               NAME                     ATTEMPT             POD ID
33e2f0bb042d0       148c192562719       23 minutes ago      Running             local-path-provisioner   0                   ebc16dddba698
95c3a52eda28f       65b87ee8c63ed       45 hours ago        Running             cert-manager             0                   42128b2ddfc32
364448138dc52       f55b2545c435d       46 hours ago        Running             minio                    0                   21b008e2f2148
7b038676524be       897ce3c5fc8ff       4 days ago          Running             lb-port-443              0                   31367827a18aa
1d61d0fad662e       897ce3c5fc8ff       4 days ago          Running             lb-port-80               0                   31367827a18aa
```
**N.B.** This issue existed only on the worker nodes and the command executed without issue on the master.

**Comment by flixr** [comment]
Should we reopen this? With v1.24.6+k3s1 it is not picking up the configured data-dir and you have to specify the config file manually.
Same goes if called via `k3s crictl`

**Comment by brandond** [comment]
@flixr I think what you're noticing is a duplicate of a different issue:
* https://github.com/k3s-io/k3s/issues/6204
